### PR TITLE
GH-98897: fix memory leak if `math.dist` raises exception

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-98897

### CODE REVIEW

**Review by vstinner — COMMENTED** [review]
Can you please add a test to make sure that we don't reintroduce this memory leak later?

**Comment by kumaraditya303** [comment]
> Can you please add a test to make sure that we don't reintroduce this memory leak later?

Added test.

File: Lib/test/test_math.py [hunk 1]
```diff
@@ -1006,6 +1007,21 @@ class T(tuple):
...
+        # See https://github.com/python/cpython/issues/98897
+        code = textwrap.dedent("""
+        import math
+        try:
+            math.dist([1, 2], [3, 4, 5])
+        except ValueError:
+            pass
+        """)
+        res = script_helper.assert_python_ok('-X', 'showrefcount', '-c', code)
+        self.assertEqual(res.err, b'[0 refs, 0 blocks]\n')
```
Comment (vstinner) [inline Lib/test/test_math.py]: That's overkill. Python test suite has a built-in memory leak detected. No need to go throw a subprocess. Trying to run ``./python -m test test_math -m test_math_dist_leak -R 3:3`` without your fix.

I suggest:

```
with self.assertRaises(ValueError):
    math.dist([1, 2], [3, 4, 5])
```

File: Lib/test/test_math.py [hunk 2]
```diff
@@ -1006,6 +1007,21 @@ class T(tuple):
             self.assertEqual(math.dist(p, q), 5*scale)
             self.assertEqual(math.dist(q, p), 5*scale)
 
+    @support.cpython_only
+    @unittest.skipUnless(support.Py_DEBUG,
+                         '-X showrefcount requires a Python debug build')
+    def test_math_dist_leak(self):
+        # See https://github.com/python/cpython/issues/98897
```
Comment (vstinner) [inline Lib/test/test_math.py]:
```suggestion
        # gh-98897: Check for error handling does not leak memory
```

File: Misc/NEWS.d/next/Library/2022-10-31-12-34-03.gh-issue-98897.rgNn4x.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix memory leak in :func:`math.dist` when it raises exception. Patch by Kumar Aditya.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2022-10-31-12-34-03.gh-issue-98897.rgNn4x.rst]:
```suggestion
Fix memory leak in :func:`math.dist` when both points don't have the same dimension. Patch by Kumar Aditya.
```

I don't think that it's useful to mention the fix on the unlikely MemoryError code path.

File: Lib/test/test_math.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/test/test_math.py]: Right, I forgot about the built-in one :)

**Review by rhettinger — APPROVED** [review]

**Comment by rhettinger** [comment]
Looks good. Thanks for the bug report and patch.

**Pull request merged**